### ISSUE TITLE
can't work

### ISSUE THREAD

**Issue by mcheung63** [pohmelie/run-in-terminal#10]
hi, when i right click the tree and click "start terminal here", nothing happen. I am in mac, how to config this plugin? thanks

**Comment by pohmelie** [comment]
Read description, there is some recepie for mac users. https://atom.io/packages/run-in-terminal

**Comment by DanHickstein** [comment]
If I add `open -a /Applications/Utilities/Terminal.app` to the "Terminal with arguments

**Comment by pohmelie** [comment]
I have just some test on my hackintosh virtualbox machine with last atom and run-in-terminal versions, and terminal starts on F9 keypress, when some file is opened and with context menu too. Please, describe more precisely your steps.

**Comment by DanHickstein** [comment]
I just added `open -a /Applications/Utilities/Terminal.app` to the "terminal with arguments" section. F9 works fine to open the Terminal.app, but I would like to be able to actually run a script with F5.

**Comment by pohmelie** [comment]
@DanHickstein, right now I'm far away from my virtualbox hackintosh and can't check arguments.
You should find documentation, or just try to see terminal `--help` for something like `working directory` and `run command` flags. Then you should paste `run command` flag to `Terminal execution argument` in run-in-terminal options and add `working directory` to `Terminal with arguments`. You will have something like `open -a /Applications/Utilities/Terminal.app (working directory flag) {working_directory}` in your `Terminal with arguments`.
To run python files you can specify shebang and check `Use shebang if available` or specify `List of launchers by extension` with `.py python3 {file_path}`

**Comment by DominikDitoIvosevic** [comment]
It doesn't work on win10 and atom 1.8.0-beta3.
None of the commands or f9 do anything

**Comment by DominikDitoIvosevic** [comment]
Yeah after reading the docs for 15min I still don't understand what do I have to do to get this running...
Could you please add steps for first time usage on windows?

**Comment by pohmelie** [comment]
@DominikDitoIvosevic, show your run-in-terminal settings.

**Comment by Kee-Wang** [comment]
I have the same issue! I'm currently using Atom 1.11.2.
I can open terminal using F5 or F9, and I am in the working directory.  But nothing happened after that. After playing this for hours (that's really frustrating man), I'm sure it is irresponsive to 'List of launchers by extension'.


'Terminal with argument':
open -a /Applications/Utilities/Terminal.app {working_directory} 

'Terminal execution argument':
-e

'List of launchers by extension': (I have tried following)
.py python {file_path}  (I don't understand this syntax but I tried anyway)
 /usr/bin/python {file_path}  (it works if I type manually in newly opened terminal)
 /usr/bin/python ./test.py     (I tried to put the path explicitly. It works if I type manually in newly opened terminal)

**Comment by pohmelie** [comment]
It looks like there is nothing wrong with "run in terminal"? am i right?

**Comment by Kee-Wang** [comment]
@pohmelie 
Yes, you are right, 'run-in-terminal' is perfectly fine. Admittedly it took me lots of time to understand what you really mean. Moreover, README.md was not easy for me to digest because konsole --noclose --workdir won't work for my Mac and worsen the situation.

**Comment by pohmelie** [comment]
Feel free to make pull to readme.md, I tried to make it friendly at my best… but probably it is still poor.

**Comment by pohmelie** [comment]
Or you mean that -e wasn't in readme?

**Comment by Kee-Wang** [comment]
@pohmelie 
I made a pull to give a full example of my case.

**Comment by jnelissen** [comment]
Hi all,

I struggled a bit as well to get things working on my mac and have another working solution which doesn't need you to chmod +x each file you want to test. Although the example from @Kee-Wang works, it is a bit tedious to chmod +x each file you want to test. The reason it works when you set the file properties to executable and add a shebang (e.g. #!/usr/bin/pyhton) in the file (but not use it in the settings!) is that the script itself is actually opened in the terminal (e.g. foo.py") instead of  "python foo.py". Bash or whatever you're using will just open foo.py and start python as that's what's found in the shebang. 
I tested this by logging the command that's actually used by child_process.exec to the console in atom. I also found that it's not possible to use the "open -a /Applications/Utilities/Terminal.app" for this purpose as it doesn't understand the arguments. 

I have used the following (somewhat ugly) setup for opening (in this case python) a script with the correct editor:
- switch Autoquote off
- In the list of launchers type:   .py python {file_path}"'
  (that's a double quote and a single quote at the end)
- In the Terminal with arguments field type: osascript -e 'tell application "Terminal" to do script "
  (doublequote at the end)
- In terminal execution argument type: ''
  (that's two single quotes, if I leave it blank the term "terminal-execution-argument" is added to my command!) This seems a bug.
- use exec cwd argument and use shebang are both switched off in this setup.

If you now press F5 or select "run-in-terminal:editor-start-terminal-here-and-run" the terminal will start (in the background) and open the file the correct way. The complete command now looks like:
osascript -e 'tell application "Terminal" to do script "python /Users/johndoe/foo.py"'

If you want to open the terminal in the foreground you can do this:

- In the list of launchers type:   .py python {file_path}"' -e 'end tell'
- In the Terminal with arguments field type:  osascript -e 'tell application "Terminal"' -e 'activate' -e 'tell application "Terminal" to do script "

If you press F5 now the terminal will start in the foreground and run python with the file as argument . Both solutions use applescript to open the terminal. This has been tested on OSX Sierra by the way. The complete command now looks like:
osascript -e 'tell application "Terminal"' -e 'activate' -e 'tell application "Terminal" to do script " python /Users/johndoe/foo.py"' -e 'end tell'

All the different quotes and workarounds to get them in the right place is a bit ugly I think, but it works for me for now.

**Comment by pohmelie** [comment]
@jnelissen 
Great, that you have success with mac, since I have no mac and can't dig into :clap: 
So, what is your suggestion? Add your solution to description or change something in package?

> if I leave it blank the term "terminal-execution-argument" is added to my command!) This seems a bug.

It's not a bug, it is default value for this field. I don't know any terminal app, which have no such flag. So it is by design right now. Maybe you can set it as space.

**Comment by jnelissen** [comment]
@pohmelie 
Since I can not think of another way to start the terminal on osx and run a command like this, the best way might be to add a toggle button specifically for Mac users that sets this command and all the strange quotes in the background of the script. If you agree I can make a start with this in the software, but I might need some help there as I'm not an expert.

Regarding the argument field, I understand the use of arguments in terminal apps, but I don't think that if you leave the field blank, the argument itself should be: "terminal-execution-argument". It should just be empty if you give no arguments right? I have changed the code from:

cmd.push(read_option("terminal_exec_arg"))
to:
xarg = read_option("terminal_exec_arg")
if xarg != "terminal-execution-argument"
   cmd.push(read_option("terminal_exec_arg"))

And now it works for me. If a add an argument e.g. -e it is loaded, but if I leave the field to the default value, no argument is given instead of the default value from the settings file.

**Comment by pohmelie** [comment]
@jnelissen 
Well, in this case we can detect os on package load and set defaults by three type of systems:
* mac — your setup
* win — well known, working and documented setup
* linux — defaults as it is now (since linux terminals is big zoo)
Detection can be done via [os.platform](https://nodejs.org/api/os.html#os_os_platform) method.

And yes, maybe we should change defaults to empty strings.

What do you think?

**Comment by jnelissen** [comment]
@pohmelie 
I'm working on the solution for mac based on your suggestion to use os.platform. Basically it will ignore the "Terminal with arguments" field and insert the correct code in the background. Setting arguments will not be useful on a mac as they won't be interpreted the correct way due to the use of applescript. I can however now set a command in the arguments field e.g. "cd {working_directory;" which will then be executed before the script runs. This is useful as the exec_cwd setting does not work in my setup either.

I'm also trying to implement the start-terminal-here without commands. I'll look into that tomorrow. 

How is F6 run-command-with-arguments supposed to work? I tried several things, but nothing seems to happen, and nothing is logged to my console.

Not sure if the option is needed for windows as well, if it is already possible to start the command line with the command in the example?

The default strings are fine I think , and helpful as they provide an example. so I wouldn't leave them blank. I made a small adjustment to not include the default value if no arguments are filled in. For the other fields it is not applicable, as there will always be some command to work on other systems.

Once I have the code ready (and figured out how Github works), I'll copy my changes there for review.
